### PR TITLE
[#61]: evPushSender can be null

### LINKED ISSUE #61: GitHub sometimes sends a pull event without a sender
**Operating system or device, package version, compiler version:**
Affects all.

**Issue description:**
In the `push` event, we were expecting the sender of the push event to always be a known User, but in some cases the User can be null.
Reported by @roberth. 

**Steps to reproduce:**
See this example fixture:
```
{
  "ref": "refs/heads/gh-readonly-queue/main/pr-3072-287d30540ac5a1acc57d0bdc138fa81dee315f48",
  "before": "a6bd77bbe77ee6ae8f9686621b192d70fcf83e2b",
  "after": "0000000000000000000000000000000000000000",
  "repository": {
    "id": 353322834,
    "node_id": "MDEwOJllG9zcXaRcvknzNYyTzM4OUQD=",
    "name": "infra-core",
    "full_name": "some-organization/infra-core",
    "private": true,
    "owner": {
      "name": "some-organization",
      "email": null,
      "login": "some-organization",
      "id": 52709322,
      "node_id": "MDEyO9ky2ZuaXpFdhlGvjbzNzczMTO3Y",
      "avatar_url": "https://avatars.githubusercontent.com/u/52709322?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/some-organization",
      "html_url": "https://github.com/some-organization",
      "followers_url": "https://api.github.com/users/some-organization/followers",
      "following_url": "https://api.github.com/users/some-organization/following{/other_user}",
      "gists_url": "https://api.github.com/users/some-organization/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/some-organization/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/some-organization/subscriptions",
      "organizations_url": "https://api.github.com/users/some-organization/orgs",
      "repos_url": "https://api.github.com/users/some-organization/repos",
      "events_url": "https://api.github.com/users/some-organization/events{/privacy}",
      "received_events_url": "https://api.github.com/users/some-organization/received_events",
      "type": "Organization",
      "site_admin": false
    },
    "html_url": "https://github.com/some-organization/infra-core",
    "description": "Homo sapiens non urinat in ventum",
    "fork": false,
    "url": "https://github.com/some-organization/infra-core",
    "forks_url": "https://api.github.com/repos/some-organization/infra-core/forks",
    "keys_url": "https://api.github.com/repos/some-organization/infra-core/keys{/key_id}",
    "collaborators_url": "https://api.github.com/repos/some-organization/infra-core/collaborators{/collaborator}",
    "teams_url": "https://api.github.com/repos/some-organization/infra-core/teams",
    "hooks_url": "https://api.github.com/repos/some-organization/infra-core/hooks",
    "issue_events_url": "https://api.github.com/repos/some-organization/infra-core/issues/events{/number}",
    "events_url": "https://api.github.com/repos/some-organization/infra-core/events",
    "assignees_url": "https://api.github.com/repos/some-organization/infra-core/assignees{/user}",
    "branches_url": "https://api.github.com/repos/some-organization/infra-core/branches{/branch}",
    "tags_url": "https://api.github.com/repos/some-organization/infra-core/tags",
    "blobs_url": "https://api.github.com/repos/some-organization/infra-core/git/blobs{/sha}",
    "git_tags_url": "https://api.github.com/repos/some-organization/infra-core/git/tags{/sha}",
    "git_refs_url": "https://api.github.com/repos/some-organization/infra-core/git/refs{/sha}",
    "trees_url": "https://api.github.com/repos/some-organization/infra-core/git/trees{/sha}",
    "statuses_url": "https://api.github.com/repos/some-organization/infra-core/statuses/{sha}",
    "languages_url": "https://api.github.com/repos/some-organization/infra-core/languages",
    "stargazers_url": "https://api.github.com/repos/some-organization/infra-core/stargazers",
    "contributors_url": "https://api.github.com/repos/some-organization/infra-core/contributors",
    "subscribers_url": "https://api.github.com/repos/some-organization/infra-core/subscribers",
    "subscription_url": "https://api.github.com/repos/some-organization/infra-core/subscription",
    "commits_url": "https://api.github.com/repos/some-organization/infra-core/commits{/sha}",
    "git_commits_url": "https://api.github.com/repos/some-organization/infra-core/git/commits{/sha}",
    "comments_url": "https://api.github.com/repos/some-organization/infra-core/comments{/number}",
    "issue_comment_url": "https://api.github.com/repos/some-organization/infra-core/issues/comments{/number}",
    "contents_url": "https://api.github.com/repos/some-organization/infra-core/contents/{+path}",
    "compare_url": "https://api.github.com/repos/some-organization/infra-core/compare/{base}...{head}",
    "merges_url": "https://api.github.com/repos/some-organization/infra-core/merges",
    "archive_url": "https://api.github.com/repos/some-organization/infra-core/{archive_format}{/ref}",
    "downloads_url": "https://api.github.com/repos/some-organization/infra-core/downloads",
    "issues_url": "https://api.github.com/repos/some-organization/infra-core/issues{/number}",
    "pulls_url": "https://api.github.com/repos/some-organization/infra-core/pulls{/number}",
    "milestones_url": "https://api.github.com/repos/some-organization/infra-core/milestones{/number}",
    "notifications_url": "https://api.github.com/repos/some-organization/infra-core/notifications{?since,all,participating}",
    "labels_url": "https://api.github.com/repos/some-organization/infra-core/labels{/name}",
    "releases_url": "https://api.github.com/repos/some-organization/infra-core/releases{/id}",
    "deployments_url": "https://api.github.com/repos/some-organization/infra-core/deployments",
    "created_at": 1618165822,
    "updated_at": "2022-11-07T17:04:28Z",
    "pushed_at": 1670262117,
    "git_url": "git://github.com/some-organization/infra-core.git",
    "ssh_url": "git@github.com:some-organization/infra-core.git",
    "clone_url": "https://github.com/some-organization/infra-core.git",
    "svn_url": "https://github.com/some-organization/infra-core",
    "homepage": "https://github.com/orgs/some-organization/projects/7",
    "size": 23932,
    "stargazers_count": 4,
    "watchers_count": 4,
    "language": "Brainfuck",
    "has_issues": true,
    "has_projects": true,
    "has_downloads": true,
    "has_wiki": true,
    "has_pages": false,
    "has_discussions": true,
    "forks_count": 1,
    "mirror_url": null,
    "archived": false,
    "disabled": false,
    "open_issues_count": 130,
    "license": {
      "key": "other",
      "name": "Other",
      "spdx_id": "NOASSERTION",
      "url": null,
      "node_id": "MDc6GTjlW5ZZzAT="
    },
    "allow_forking": false,
    "is_template": false,
    "web_commit_signoff_required": false,
    "topics": [

    ],
    "visibility": "private",
    "forks": 1,
    "open_issues": 130,
    "watchers": 4,
    "default_branch": "main",
    "stargazers": 4,
    "master_branch": "main",
    "organization": "some-organization"
  },
  "pusher": {
    "name": "none"
  },
  "organization": {
    "login": "some-organization",
    "id": 52709322,
    "node_id": "MDEyO9kZyFua2pXhGlvdjbzNcMzOTYz3",
    "url": "https://api.github.com/orgs/some-organization",
    "repos_url": "https://api.github.com/orgs/some-organization/repos",
    "events_url": "https://api.github.com/orgs/some-organization/events",
    "hooks_url": "https://api.github.com/orgs/some-organization/hooks",
    "issues_url": "https://api.github.com/orgs/some-organization/issues",
    "members_url": "https://api.github.com/orgs/some-organization/members{/member}",
    "public_members_url": "https://api.github.com/orgs/some-organization/public_members{/member}",
    "avatar_url": "https://avatars.githubusercontent.com/u/52709322?v=4",
    "description": "Plurality Media"
  },
  "installation": {
    "id": 17816238,
    "node_id": "MDIzklOdGunVcFm0WauSW9z5GFdsGb0FW9uaTMgM3zcNjg4="
  },
  "created": false,
  "deleted": true,
  "forced": false,
  "base_ref": null,
  "compare": "https://github.com/some-organization/infra-core/compare/a6bd77bbe77e...000000000000",
  "commits": [

  ],
  "head_commit": null
}
```

**Link to minimal example project:**
N/A

### PR DESCRIPTION
**Pull request by roberth**
**Issue reference:**

Fixes #61 
GitHub sometimes sends a `pull` event without a sender. See the anonymized test case.

**Submission Checklist:**
* [x] Have you followed the guidelines in our [Contributing](../../CONTRIBUTING.md) document (for example, is your tree a clean merge)?
  * Broken link
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
* [x] Does your submission build?
  * built on linux with haskell.nix
* [x] Does your submission pass tests?
* [x] Have you run lints on your code locally prior to submission?
* [x] Have you updated *all* of the cabal/nix infrastructure?
* [x] Is this a breaking change? Have you discussed this?
  * Formally it is a breaking change, as `PushEvent` consumers may have to handle a `Nothing`, but this merely reflects the reality of github webhooks. `sender` may be omitted in rare cases.

### CODE REVIEW

**Comment by kvanbere** [comment]
Hi Robert, thanks for the PR!

I'm currently away, but I'll take a look and get back to you within the next day or so.

**Review by kvanbere — APPROVED** [review]

**Comment by kvanbere** [comment]
@roberth Changes look good, but I want to run the CI over this. I have to make a few changes so that it can be run on a remote fork since the GitHub `push` event does not trigger due to the way this was raised.

File: spec/DecodeEventsSpec.hs [hunk 1]
```diff
@@ -5653,6 +5654,129 @@ pushEventFixture = PushEvent
...
+  PushEvent
+    { evPushRef = "refs/heads/gh-readonly-queue/main/pr-3072-287d30540ac5a1acc57d0bdc138fa81dee315f48",
+      evPushHeadSha = Just "0000000000000000000000000000000000000000",
+      evPushBeforeSha = Just "a6bd77bbe77ee6ae8f9686621b192d70fcf83e2b",
+      evPushCreated = False,
+      evPushDeleted = True,
+      evPushForced = False,
+      evPushBaseRef = Nothing,
+      evPushCompareUrl = URL "https://github.com/some-organization/infra-core/compare/a6bd77bbe77e...000000000000",
+      evPushCommits = Just mempty,
```
Comment (kvanbere) [inline spec/DecodeEventsSpec.hs]: Please see: https://github.com/cuedo/github-webhooks/actions/runs/3680596117/jobs/6226368854#step:4:172

**Review by kvanbere — CHANGES_REQUESTED** [review]
See inline comment on CI failure for `mempty`.

File: spec/DecodeEventsSpec.hs (hunk 1 above)
Comment (roberth) [reply spec/DecodeEventsSpec.hs]: 7.8...
I'm going to fix it now, but are you sure you need a 7 year old compiler?

Comment (kvanbere) [reply spec/DecodeEventsSpec.hs]: The fix is simple, so let's keep compatibility for now. Thanks

**Review by kvanbere — CHANGES_REQUESTED** [review]
Looks like the CI has raised warnings that we forgot to export `EventHasMaybeSender(..)`.

File: spec/DecodeEventsSpec.hs (hunk 1 above)
Comment (roberth) [reply spec/DecodeEventsSpec.hs]: Fair enough, appreciate the commitment.

**Review by kvanbere — CHANGES_REQUESTED** [review]
Hi Rob, thanks for that revision. According to CI, it looks like one of the examples (`github-webhooks-servant-example`) needs to be updated to match this change to the API:
```
2022-12-13T00:51:21.1290270Z github-webhooks-servant-example       > Configuring github-webhooks-servant-example-0.0.0...
2022-12-13T00:51:21.3253306Z github-webhooks-servant-simple-example> configure (exe)
2022-12-13T00:51:21.3257387Z github-webhooks-servant-example       > build (exe)
2022-12-13T00:51:21.4035215Z github-webhooks-servant-example       > Preprocessing executable 'github-webhooks-servant-example' for github-webhooks-servant-example-0.0.0..
2022-12-13T00:51:21.4036052Z github-webhooks-servant-example       > Building executable 'github-webhooks-servant-example' for github-webhooks-servant-example-0.0.0..
2022-12-13T00:51:21.5255894Z github-webhooks-servant-example       > [1 of 2] Compiling Main
2022-12-13T00:51:21.9723780Z github-webhooks-servant-simple-example> Configuring github-webhooks-servant-simple-example-0.0.0...
2022-12-13T00:51:22.0500611Z github-webhooks-servant-example       > 
2022-12-13T00:51:22.0502246Z github-webhooks-servant-example       > /home/runner/work/github-webhooks/github-webhooks/examples/servant/src/Main.hs:38:38: error:
2022-12-13T00:51:22.0502986Z github-webhooks-servant-example       >     • Couldn't match type ‘Maybe HookUser’ with ‘HookUser’
2022-12-13T00:51:22.0503531Z github-webhooks-servant-example       >       Expected type: PushEvent -> HookUser
2022-12-13T00:51:22.0504386Z github-webhooks-servant-example       >         Actual type: PushEvent -> Maybe HookUser
2022-12-13T00:51:22.0504908Z github-webhooks-servant-example       >     • In the second argument of ‘(.)’, namely ‘evPushSender’
2022-12-13T00:51:22.0505409Z github-webhooks-servant-example       >       In the second argument of ‘(.)’, namely
2022-12-13T00:51:22.0505871Z github-webhooks-servant-example       >         ‘whUserLogin . evPushSender’
2022-12-13T00:51:22.0506373Z github-webhooks-servant-example       >       In the expression: show . whUserLogin . evPushSender
2022-12-13T00:51:22.0506787Z github-webhooks-servant-example       >    |
2022-12-13T00:51:22.0507361Z github-webhooks-servant-example       > 38 |     putStrLn $ (show . whUserLogin . evPushSender) ev ++ " pushed a commit causing HEAD SHA to become:"
2022-12-13T00:51:22.0507911Z github-webhooks-servant-example       >    |  
```
https://github.com/cuedo/github-webhooks/actions/runs/3681095902/jobs/6227414754

The examples are designed to be overly-simple, so I think using `fromJust` would be fine here.

**Comment by roberth** [comment]
Thanks for helping!

**Pull request merged**